tests/e2e/e2e_garbage_collector_test.go

```diff
@@ -0,0 +1,135 @@
+			userReplicaSet := 5
+
+			var userServices []string
+
+			for i := 0; i < numberOfUserServices; i++ {
+				userServices = append(userServices, fmt.Sprintf("%s%d", userAppBaseName, i))
+			}
+
+			// Used across the test to wait for concurrent steps to finish
+			var wg sync.WaitGroup
+
+			It("Tests garbage collections", func() {
+				// Install OSM
+				Expect(Td.InstallOSM(Td.GetOSMInstallOpts())).To(Succeed())
+
+				// Create NSs
+				Expect(Td.CreateMultipleNs(userServices...)).To(Succeed())
+				Expect(Td.AddNsToMesh(true, userServices...)).To(Succeed())
+
+				// User apps
+				for _, userApp := range userServices {
+					svcAccDef, deploymentDef, svcDef := Td.SimpleDeploymentApp(
+						SimpleDeploymentAppDef{
+							Name:         userApp,
+							Namespace:    userApp,
+							ReplicaCount: int32(userReplicaSet),
+							Command:      []string{"/bin/bash", "-c", "--"},
+							Args:         []string{"while true; do sleep 30; done;"},
+							Image:        "songrgg/alpine-debug",
+							Ports:        []int{80},
+						})
+
+					_, err := Td.CreateServiceAccount(userApp, &svcAccDef)
+					Expect(err).NotTo(HaveOccurred())
+					_, err = Td.CreateDeployment(userApp, deploymentDef)
+					Expect(err).NotTo(HaveOccurred())
+					_, err = Td.CreateService(userApp, svcDef)
+					Expect(err).NotTo(HaveOccurred())
+
+					wg.Add(1)
+					go func(app string) {
+						defer GinkgoRecover()
+						defer wg.Done()
+						Expect(Td.WaitForPodsRunningReady(app, 200*time.Second, userReplicaSet)).To(Succeed())
+					}(userApp)
+				}
+
+				wg.Wait()
+
+				By("Verifying the secrets have been patched with OwnerReference")
+
+				podSelector := constants.EnvoyUniqueIDLabelName
+				for _, ns := range userServices {
+					pods, err := Td.Client.CoreV1().Pods(ns).List(context.Background(), metav1.ListOptions{LabelSelector: podSelector})
+					Expect(err).To(BeNil())
+
+					for _, pod := range pods.Items {
+						podUUID := pod.GetLabels()[podSelector]
+						secretSelector := fmt.Sprintf("%s=%s", constants.EnvoyUniqueIDLabelName, podUUID)
+						secretList, err := Td.Client.CoreV1().Secrets(ns).List(context.Background(), metav1.ListOptions{LabelSelector: secretSelector})
+						Expect(err).To(BeNil())
+						Expect(len(secretList.Items)).To(BeEquivalentTo(1))
+
+						secret := secretList.Items[0]
+						ownerReferences := secret.GetOwnerReferences()
+						Expect(ownerReferences).ToNot(BeNil())
+
+						expectedOwnerReference := v1.OwnerReference{
+							APIVersion: "v1",
+							Kind:       "Pod",
+							Name:       pod.GetName(),
+							UID:        pod.GetUID(),
+						}
+
+						foundReference := false
+						for _, ownerReference := range ownerReferences {
+							if reflect.DeepEqual(expectedOwnerReference, ownerReference) {
+								foundReference = true
+							}
+						}
+						Expect(foundReference).To(BeTrue())
+					}
+				}
+
+				By("Verifying unused secrets are deleted when the referenced owner is deleted")
```

Comment:
This seems to be creeping into "testing Kubernetes" territory which we shouldn't feel a need to do, but since it's the only reason we added this in the first place I'd say it's good to keep.